smbclient/shutil.py

```diff
@@ -0,0 +1,306 @@
+    :return: True if path is a dir or points to a dir.
+    """
+    if ignore_errors:
+        def onerror(*args):
+            pass
+    elif onerror is None:
+        def onerror(*args):
+            raise
+    try:
+        if islink(path, **kwargs):
+            # symlinks to directories are forbidden, see bug #1669
+            raise OSError("Cannot call rmtree on a symbolic link")
+    except OSError:
+        onerror(islink, path, sys.exc_info())
+        # can't continue even if onerror hook returns
+        return
+
+    names = []
+    try:
+        names = listdir(path, **kwargs)
+    except os_error:
+        onerror(listdir, path, sys.exc_info())
+    for name in names:
+        fullname = join(path, name)
+        try:
+            mode = lstat(fullname, **kwargs).st_mode
+        except os_error:
+            mode = 0
+        if py_stat.S_ISDIR(mode):
+            rmtree(fullname, ignore_errors, onerror, **kwargs)
+        else:
+            try:
+                remove(fullname, **kwargs)
+            except os_error:
+                onerror(remove, fullname, sys.exc_info())
+    try:
+        rmdir(path, **kwargs)
+    except os_error:
+        onerror(rmdir, path, sys.exc_info())
+
+
+def copy2(src, dst, server_side_copy=False, **kwargs):
+    """Copy data and metadata. Return the file's destination.
+
+    Metadata is copied with copystat(). Please see the copystat function
+    for more information.
+
+    The destination may be a directory.
+
+    """
+    if isdir(dst, **kwargs):
+        dst = join(dst, basename(src))
+    if server_side_copy:
+        copyfile_server_side(src, dst, **kwargs)
+    else:
+        copyfile(src, dst, **kwargs)
+    copystat(src, dst, **kwargs)
+
+
+def copystat(src, dst, **kwargs):
+    """Copy file metadata
+
+    Copy the permission bits, last access time, last modification time, and
+    flags from `src` to `dst`. The file contents, owner, and group are
+    unaffected. `src` and `dst` are path names given as strings.
+    """
+    st = stat(src, **kwargs)
+    utime(dst, ns=(st.st_atime_ns, st.st_mtime_ns), **kwargs)
+
+    # TODO: in principle, we can copy the readonly flag via chmod
+    # mode = py_stat.S_IMODE(st.st_mode)
+    # if hasattr(os, 'chmod'):
+    #     os.chmod(dst, mode)
+
+
+def copyfile(src, dst, **kwargs):
+    """Copy data from src to dst"""
+    if _samefile(src, dst):
+        raise Error("`%s` and `%s` are the same file" % (src, dst))
+
+    for fn in [src, dst]:
+        try:
+            st = stat(fn, **kwargs)
+        except OSError:
+            # File most likely does not exist
+            pass
+        else:
+            # XXX What about other special files? (sockets, devices...)
+            if py_stat.S_ISFIFO(st.st_mode):
+                raise SpecialFileError("`%s` is a named pipe" % fn)
+
+    with open_file(src, 'rb', **kwargs) as fsrc:
+        with open_file(dst, 'wb', **kwargs) as fdst:
```

Comment:
`with` supports multiple entries so this could just be `with open_file(src, 'rb', **kwargs) as fsrc, open_file(dst, 'wb', **kwargs) as fdst:`. https://docs.python.org/3/reference/compound_stmts.html#the-with-statement

Reply:
non server side copy code removed for now